File: arrays/firstBadVersion.go
package main


/*

link problem leetcode -> https://leetcode.com/problems/first-bad-version
time complexity  -> O(log(n))
space complexity -> O(1)

*/
func isBadVersion(version int) bool;

func firstBadVersion(n int) int {
	l,r := 1,n
	middle := 0
	for l < r {
		
		middle = (l + r) / 2
		if isBadVersion(middle){
			r = middle
		}else{
			l = middle + 1
		}
	}  
 
	return l
 }

File: binary_search/first_bad_version.py
#Link to problem: https://leetcode.com/problems/first-bad-version/description/
#Time complexity: O(log(n))
#Space complexity: O(1)

from help.isBadVersion_definition import isBadVersion

class Solution:
    def first_bad_version(self,n:int) -> int:
        left = 0
        right = n - 1

        while left <= right:

            mid = (left + right) // 2

            if isBadVersion(mid):
                right = mid - 1
            else:
                left = mid + 1

        return left + 1
                                        
solution = Solution()
print(solution.first_bad_version(5))
#print(solution.first_bad_version(1))
